app/src/main/java/com/zeapo/pwdstore/git/config/SshjConfig.kt

```diff
@@ -0,0 +1,142 @@
+import net.schmizz.sshj.common.SecurityUtils
+import net.schmizz.sshj.transport.compression.NoneCompression
+import net.schmizz.sshj.transport.kex.Curve25519SHA256
+import net.schmizz.sshj.transport.kex.Curve25519SHA256.FactoryLibSsh
+import net.schmizz.sshj.transport.kex.DHGexSHA256
+import net.schmizz.sshj.transport.kex.ECDHNistP
+import net.schmizz.sshj.transport.random.BouncyCastleRandom
+import net.schmizz.sshj.transport.random.SingletonRandomFactory
+import net.schmizz.sshj.userauth.keyprovider.OpenSSHKeyFile
+import net.schmizz.sshj.userauth.keyprovider.PKCS5KeyFile
+import net.schmizz.sshj.userauth.keyprovider.PKCS8KeyFile
+import net.schmizz.sshj.userauth.keyprovider.PuTTYKeyFile
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import java.security.Security
+
+
+fun setUpBouncyCastleForSshj() {
+    // Replace the Android BC provider with the Java BouncyCastle provider since the former does
+    // not include all the required algorithms.
+    // TODO: Verify that we are indeed using the fast Android-native implementation whenever
+    //  possible.
+    // Note: This may affect crypto operations in other parts of the application.
+    val bcIndex = Security.getProviders().indexOfFirst {
+        it.name == BouncyCastleProvider.PROVIDER_NAME
+    }
+    if (bcIndex == -1) {
+        // No Android BC found, install Java BC at lowest priority.
+        Security.addProvider(BouncyCastleProvider())
+    } else {
+        // Replace Android BC with Java BC, inserted at the same position.
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME)
+        // May be needed on Android Pie+ as per https://stackoverflow.com/a/57897224/297261
+        try {
+            Class.forName("sun.security.jca.Providers")
+        } catch (e: ClassNotFoundException) {
+        }
+        Security.insertProviderAt(BouncyCastleProvider(), bcIndex + 1)
+    }
+    d { "JCE providers: ${Security.getProviders().joinToString { "${it.name} (${it.version})" }}" }
+    // By setting this to false, we prevent SSHJ from specifying the BC provider explicitly, which
+    // is preferred on Android and ensures compatibility with Android Keystore keys.
+    SecurityUtils.setRegisterBouncyCastle(false)
+    SecurityUtils.setSecurityProvider(null)
+}
+
+/**
+ * Based on SSHJ's DefaultConfig.java, which is licensed under the Apache License, Version 2.0, with
+ * the following copyright notice:
+ * Copyright (C)2009 - SSHJ Contributors
+ */
+class SshjConfig : ConfigImpl() {
+
+    init {
+        // TODO: Replace with a Timber-backed LoggerFactory
+        loggerFactory = LoggerFactory.DEFAULT
+        keepAliveProvider = KeepAliveProvider.HEARTBEAT
+
+        initKeyExchangeFactories()
+        initKeyAlgorithms()
+        initRandomFactory()
+        initFileKeyProviderFactories()
+        initCipherFactories()
+        initCompressionFactories()
+        initMACFactories()
+    }
+
+    private fun initKeyExchangeFactories() {
```

Comment:
I got rid of plenty of algorithms here compared to https://github.com/hierynomus/sshj/blob/master/src/main/java/net/schmizz/sshj/DefaultConfig.java to keep us in line with the [Mozilla Modern SSH config](https://infosec.mozilla.org/guidelines/openssh#modern). SSHJ does not have AES-GCM support (most likely a good thing given that it would have to be implemented in Java), but hopefully adopts ChaCha20-Poly1305 at some point.

Reply:
I looked at the Mozilla config and I think it's a pretty sane collection, we should be fine. Would certainly love to have ChaCha20.

Reply:
Unfortunately all the ciphers currently supported by SSHJ are unauthenticated. Getting ChaCha20-Poly1305 support in there would require some structural changes and thus support by the current maintainer. We should definitely revisit that point in the future.